flang/lib/Optimizer/Support/DataLayout.cpp

```diff
@@ -46,15 +49,53 @@ void fir::support::setMLIRDataLayoutFromAttributes(mlir::ModuleOp mlirModule,
   fir::support::setMLIRDataLayout(mlirModule, llvmDataLayout);
 }
 
-std::optional<mlir::DataLayout>
-fir::support::getOrSetDataLayout(mlir::ModuleOp mlirModule,
-                                 bool allowDefaultLayout) {
+template <typename ModOpTy>
+std::optional<mlir::DataLayout> getOrSetDataLayout(ModOpTy mlirModule,
+                                                   bool allowDefaultLayout) {
   if (!mlirModule.getDataLayoutSpec()) {
     fir::support::setMLIRDataLayoutFromAttributes(mlirModule,
                                                   allowDefaultLayout);
-    if (!mlirModule.getDataLayoutSpec()) {
+    // if it is a GPU module, we let it proceed, as it's contained within
+    // a module, its parent may have a DataLayout that can take its
+    // place.
+    if (!mlirModule.getDataLayoutSpec() &&
```

Comment:
Maybe we can linearize this code instead of nesting so it is clear that the same value is checked and potentially modified.